kubectl panic on large lists 1.11.1

**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug


**What happened**:
Running `kubectl get pods --namespace=kube-system` with large amount of pods (over 100 at least? But there might be something else as it crashed after updating the deployment several times) seems to crash `kubectl`. Tested with kubectl 1.11.1 - broken, downgrading to 1.10.5 and all works fine.

**What you expected to happen**:
Not crash :thinking: 

**How to reproduce it (as minimally and precisely as possible)**:
Not sure, probably spam a lot of containers? I've linked the json it is trying to parse is uploaded here (too long to GIST): https://repo.foxiehost.eu/pods.json

**Anything else we need to know?**:
Stack trace:
```
# kubectl get pods --namespace=kube-system  
panic: runtime error: index out of range

goroutine 1 [running]:
k8s.io/kubernetes/pkg/printers.PrintTable(0xc420e5e000, 0x1893a20, 0xc4207e9c00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/printers/humanreadable.go:366 +0x5d6
k8s.io/kubernetes/pkg/printers.(*HumanReadablePrinter).PrintObj(0xc420a12300, 0x189d3c0, 0xc420e5e000, 0x1893a20, 0xc4207e9c00, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/printers/humanreadable.go:286 +0x7ca
k8s.io/kubernetes/pkg/printers.(ResourcePrinter).PrintObj-fm(0x189d3c0, 0xc420e5e000, 0x1893a20, 0xc4207e9c00, 0x1892b20, 0xc420ffe3c0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/get/get.go:252 +0x57
k8s.io/kubernetes/pkg/printers.ResourcePrinterFunc.PrintObj(0xc420bad5c0, 0x189d3c0, 0xc420e5e000, 0x1893a20, 0xc4207e9c00, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/printers/interface.go:38 +0x4e
k8s.io/kubernetes/pkg/kubectl/cmd/get.(*GetOptions).Run(0xc420336380, 0x18c6de0, 0xc42057cfc0, 0xc42029c280, 0xc4205e60c0, 0x1, 0x2, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/get/get.go:444 +0xae9
k8s.io/kubernetes/pkg/kubectl/cmd/get.NewCmdGet.func1(0xc42029c280, 0xc4205e60c0, 0x1, 0x2)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/get/get.go:159 +0x115
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute(0xc42029c280, 0xc4205e6060, 0x2, 0x2, 0xc42029c280, 0xc4205e6060)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:760 +0x2c1
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc42037b180, 0xc42026d410, 0x12a05f200, 0xc4209f5ee8)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:846 +0x30a
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute(0xc42037b180, 0x17ea8b0, 0x23ee060)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:794 +0x2b
main.main()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubectl/kubectl.go:50 +0x196
```

or if you prefer with higher verbosity
```
# kubectl get pods --namespace=kube-system   --v=8
I0719 14:12:30.192831   34974 loader.go:359] Config loaded from file /home/ashley/.kube/config
I0719 14:12:30.193064   34974 loader.go:359] Config loaded from file /home/ashley/.kube/config
I0719 14:12:30.193423   34974 loader.go:359] Config loaded from file /home/ashley/.kube/config
I0719 14:12:30.198949   34974 loader.go:359] Config loaded from file /home/ashley/.kube/config
I0719 14:12:30.199210   34974 round_trippers.go:383] GET http://localhost:8080/api/v1/namespaces/kube-system/pods?limit=500
I0719 14:12:30.199223   34974 round_trippers.go:390] Request Headers:
I0719 14:12:30.199229   34974 round_trippers.go:393]     Accept: application/json;as=Table;v=v1beta1;g=meta.k8s.io, application/json
I0719 14:12:30.199234   34974 round_trippers.go:393]     User-Agent: kubectl/v1.11.1 (linux/amd64) kubernetes/b1b2997
I0719 14:12:30.209788   34974 round_trippers.go:408] Response Status: 200 OK in 10 milliseconds
I0719 14:12:30.209803   34974 round_trippers.go:411] Response Headers:
I0719 14:12:30.209808   34974 round_trippers.go:414]     Content-Type: application/json
I0719 14:12:30.209812   34974 round_trippers.go:414]     Date: Thu, 19 Jul 2018 12:12:30 GMT
I0719 14:12:30.210337   34974 request.go:897] Response Body: {"kind":"Table","apiVersion":"meta.k8s.io/v1beta1","metadata":{"selfLink":"/api/v1/namespaces/kube-system/pods","resourceVersion":"1214855237"},"columnDefinitions":[{"name":"Name","type":"string","format":"name","description":"Name must be unique within a namespace. Is required when creating resources, although some resources may allow a client to request the generation of an appropriate name automatically. Name is primarily intended for creation idempotence and configuration definition. Cannot be updated. More info: http://kubernetes.io/docs/user-guide/identifiers#names","priority":0},{"name":"Ready","type":"string","format":"","description":"The aggregate readiness state of this pod for accepting traffic.","priority":0},{"name":"Status","type":"string","format":"","description":"The aggregate status of the containers in this pod.","priority":0},{"name":"Restarts","type":"integer","format":"","description":"The number of times the containers in this pod have been restarted.","priority":0},{"name":"Age","type [truncated 98737 chars]
I0719 14:12:30.224052   34974 get.go:443] no kind is registered for the type v1beta1.Table
panic: runtime error: index out of range

goroutine 1 [running]:
k8s.io/kubernetes/pkg/printers.PrintTable(0xc420bcf280, 0x1893a20, 0xc420721b00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/printers/humanreadable.go:366 +0x5d6
k8s.io/kubernetes/pkg/printers.(*HumanReadablePrinter).PrintObj(0xc420d700c0, 0x189d3c0, 0xc420bcf280, 0x1893a20, 0xc420721b00, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/printers/humanreadable.go:286 +0x7ca
k8s.io/kubernetes/pkg/printers.(ResourcePrinter).PrintObj-fm(0x189d3c0, 0xc420bcf280, 0x1893a20, 0xc420721b00, 0x1, 0x1)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/get/get.go:252 +0x57
k8s.io/kubernetes/pkg/printers.ResourcePrinterFunc.PrintObj(0xc420ab3ec0, 0x189d3c0, 0xc420bcf280, 0x1893a20, 0xc420721b00, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/printers/interface.go:38 +0x4e
k8s.io/kubernetes/pkg/kubectl/cmd/get.(*GetOptions).Run(0xc42034c380, 0x18c6de0, 0xc42074d200, 0xc420718280, 0xc4205e7f80, 0x1, 0x3, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/get/get.go:444 +0xae9
k8s.io/kubernetes/pkg/kubectl/cmd/get.NewCmdGet.func1(0xc420718280, 0xc4205e7f80, 0x1, 0x3)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/get/get.go:159 +0x115
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute(0xc420718280, 0xc4205e7ef0, 0x3, 0x3, 0xc420718280, 0xc4205e7ef0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:760 +0x2c1
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc4201e2f00, 0xc4200fb410, 0x12a05f200, 0xc4209dfee8)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:846 +0x30a
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute(0xc4201e2f00, 0x17ea8b0, 0x23ee060)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:794 +0x2b
main.main()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubectl/kubectl.go:50 +0x196
```

**Environment**:
- Kubernetes version (use `kubectl version`): 1.11.1
- Cloud provider or hardware configuration: onprem
- OS (e.g. from /etc/os-release): debian
- Kernel (e.g. `uname -a`): debian 4.9
- Install tools: built from source

Ashley

--- a/pkg/kubectl/cmd/get/humanreadable_flags_test.go
+++ b/pkg/kubectl/cmd/get/humanreadable_flags_test.go
@@ -61,7 +61,7 @@ func TestHumanReadablePrinterSupportsExpectedOptions(t *testing.T) {
 		{
 			name:           "\"wide\" output format prints",
 			outputFormat:   "wide",
-			expectedOutput: "NAME\\ +READY\\ +STATUS\\ +RESTARTS\\ +AGE\\ +IP\\ +NODE\nfoo\\ +0/0\\ +0\\ +<unknown>\\ +<none>\\ +<none>\n",
+			expectedOutput: "NAME\\ +READY\\ +STATUS\\ +RESTARTS\\ +AGE\\ +IP\\ +NODE\\ +NOMINATED NODE\nfoo\\ +0/0\\ +0\\ +<unknown>\\ +<none>\\ +<none>\\ +<none>\n",
 		},
 		{
 			name:           "no-headers prints output with no headers",
@@ -72,7 +72,7 @@ func TestHumanReadablePrinterSupportsExpectedOptions(t *testing.T) {
 			name:           "no-headers and a \"wide\" output format prints output with no headers and additional columns",
 			outputFormat:   "wide",
 			noHeaders:      true,
-			expectedOutput: "foo\\ +0/0\\ +0\\ +<unknown>\\ +<none>\\ +<none>\n",
+			expectedOutput: "foo\\ +0/0\\ +0\\ +<unknown>\\ +<none>\\ +<none>\\ +<none>\n",
 		},
 		{
 			name:           "show-kind displays the resource's kind, even when printing a single type of resource",

--- a/pkg/printers/humanreadable.go
+++ b/pkg/printers/humanreadable.go
@@ -371,6 +371,11 @@ func PrintTable(table *metav1beta1.Table, output io.Writer, options PrintOptions
 	for _, row := range table.Rows {
 		first := true
 		for i, cell := range row.Cells {
+			if i >= len(table.ColumnDefinitions) {
+				// https://issue.k8s.io/66379
+				// don't panic in case of bad output from the server, with more cells than column definitions
+				break
+			}
 			column := table.ColumnDefinitions[i]
 			if !options.Wide && column.Priority != 0 {
 				continue

--- a/pkg/printers/internalversion/printers.go
+++ b/pkg/printers/internalversion/printers.go
@@ -84,6 +84,7 @@ func AddHandlers(h printers.PrintHandler) {
 		{Name: "Age", Type: "string", Description: metav1.ObjectMeta{}.SwaggerDoc()["creationTimestamp"]},
 		{Name: "IP", Type: "string", Priority: 1, Description: apiv1.PodStatus{}.SwaggerDoc()["podIP"]},
 		{Name: "Node", Type: "string", Priority: 1, Description: apiv1.PodSpec{}.SwaggerDoc()["nodeName"]},
+		{Name: "Nominated Node", Type: "string", Priority: 1, Description: apiv1.PodStatus{}.SwaggerDoc()["nominatedNodeName"]},
 	}
 	h.TableHandler(podColumnDefinitions, printPodList)
 	h.TableHandler(podColumnDefinitions, printPod)
@@ -632,6 +633,7 @@ func printPod(pod *api.Pod, options printers.PrintOptions) ([]metav1beta1.TableR
 
 	if options.Wide {
 		nodeName := pod.Spec.NodeName
+		nominatedNodeName := pod.Status.NominatedNodeName
 		podIP := pod.Status.PodIP
 		if podIP == "" {
 			podIP = "<none>"
@@ -639,10 +641,10 @@ func printPod(pod *api.Pod, options printers.PrintOptions) ([]metav1beta1.TableR
 		if nodeName == "" {
 			nodeName = "<none>"
 		}
-		row.Cells = append(row.Cells, podIP, nodeName)
-		if len(pod.Status.NominatedNodeName) > 0 {
-			row.Cells = append(row.Cells, pod.Status.NominatedNodeName)
+		if nominatedNodeName == "" {
+			nominatedNodeName = "<none>"
 		}
+		row.Cells = append(row.Cells, podIP, nodeName, nominatedNodeName)
 	}
 
 	return []metav1beta1.TableRow{row}, nil

--- a/pkg/printers/internalversion/printers_test.go
+++ b/pkg/printers/internalversion/printers_test.go
@@ -1686,7 +1686,7 @@ func TestPrintPodwide(t *testing.T) {
 					},
 				},
 			},
-			[]metav1beta1.TableRow{{Cells: []interface{}{"test2", "1/2", "ContainerWaitingReason", int64(6), "<unknown>", "<none>", "<none>"}}},
+			[]metav1beta1.TableRow{{Cells: []interface{}{"test2", "1/2", "ContainerWaitingReason", int64(6), "<unknown>", "<none>", "<none>", "<none>"}}},
 		},
 	}
 

--- a/pkg/registry/core/pod/storage/storage_test.go
+++ b/pkg/registry/core/pod/storage/storage_test.go
@@ -417,6 +417,7 @@ func TestConvertToTableList(t *testing.T) {
 		{Name: "Age", Type: "string", Description: metav1.ObjectMeta{}.SwaggerDoc()["creationTimestamp"]},
 		{Name: "IP", Type: "string", Priority: 1, Description: v1.PodStatus{}.SwaggerDoc()["podIP"]},
 		{Name: "Node", Type: "string", Priority: 1, Description: v1.PodSpec{}.SwaggerDoc()["nodeName"]},
+		{Name: "Nominated Node", Type: "string", Priority: 1, Description: v1.PodStatus{}.SwaggerDoc()["nominatedNodeName"]},
 	}
 
 	pod1 := &api.Pod{
@@ -435,6 +436,7 @@ func TestConvertToTableList(t *testing.T) {
 				{Name: "ctr1", State: api.ContainerState{Running: &api.ContainerStateRunning{}}, RestartCount: 10, Ready: true},
 				{Name: "ctr2", State: api.ContainerState{Waiting: &api.ContainerStateWaiting{}}, RestartCount: 0},
 			},
+			NominatedNodeName: "nominated-node",
 		},
 	}
 
@@ -452,7 +454,7 @@ func TestConvertToTableList(t *testing.T) {
 			out: &metav1beta1.Table{
 				ColumnDefinitions: columns,
 				Rows: []metav1beta1.TableRow{
-					{Cells: []interface{}{"", "0/0", "", int64(0), "<unknown>", "<none>", "<none>"}, Object: runtime.RawExtension{Object: &api.Pod{}}},
+					{Cells: []interface{}{"", "0/0", "", int64(0), "<unknown>", "<none>", "<none>", "<none>"}, Object: runtime.RawExtension{Object: &api.Pod{}}},
 				},
 			},
 		},
@@ -461,7 +463,7 @@ func TestConvertToTableList(t *testing.T) {
 			out: &metav1beta1.Table{
 				ColumnDefinitions: columns,
 				Rows: []metav1beta1.TableRow{
-					{Cells: []interface{}{"foo", "1/2", "Pending", int64(10), "1y", "10.1.2.3", "test-node"}, Object: runtime.RawExtension{Object: pod1}},
+					{Cells: []interface{}{"foo", "1/2", "Pending", int64(10), "1y", "10.1.2.3", "test-node", "nominated-node"}, Object: runtime.RawExtension{Object: pod1}},
 				},
 			},
 		},

